xdsl/dialects/builtin.py

```diff
@@ -2236,6 +2239,26 @@ def create_dense_float(
         else:
             data = cast(Sequence[float], data)
 
+        assert not isinstance(type.element_type, ComplexType)
+        return DenseIntOrFPElementsAttr([type, BytesAttr(type.element_type.pack(data))])
+
+    @overload
+    @staticmethod
+    def create_dense_complex(
+        type: RankedStructure[ComplexType], data: Sequence[tuple[int, int]]
+    ) -> DenseIntOrFPElementsAttr[ComplexType]: ...
```

Comment:
Does python let you nest generics? 
```suggestion
    ) -> DenseIntOrFPElementsAttr[ComplexType[IntegerType]]: ...
```

Reply:
I don't think so:

```
/home/ubuntu/Code/xdsl/xdsl/dialects/builtin.py
  /home/ubuntu/Code/xdsl/xdsl/dialects/builtin.py:2279:9 - error: Overload 1 for "create_dense_complex" overlaps overload 2 and returns an incompatible type (reportOverlappingOverload)
  /home/ubuntu/Code/xdsl/xdsl/dialects/builtin.py:2281:35 - error: Type "ComplexType[IntegerType]" cannot be assigned to type variable "DenseElementCovT@DenseIntOrFPElementsAttr"
    Type "ComplexType[IntegerType]" is not assignable to upper bound "AnyDenseElement" for type variable "DenseElementCovT@DenseIntOrFPElementsAttr"
      Type "ComplexType[IntegerType]" is not assignable to type "AnyDenseElement"
        "ComplexType[IntegerType]" is not assignable to "IntegerType"
        "ComplexType[IntegerType]" is not assignable to "IndexType"
        "ComplexType[IntegerType]" is not assignable to "BFloat16Type"
        "ComplexType[IntegerType]" is not assignable to "Float16Type"
        "ComplexType[IntegerType]" is not assignable to "Float32Type"
        "ComplexType[IntegerType]" is not assignable to "Float64Type"
# and others.
```

Reply:
This will work if you add `covariant = True` to `ComplexElementT`. 
You can then even do
```python
@overload
    @staticmethod
    def create_dense_complex(
        type: RankedStructure[ComplexType[_IntegerTypeInvT]], data: Sequence[tuple[int, int]]
    ) -> DenseIntOrFPElementsAttr[ComplexType[_IntegerTypeInvT]]: ...

    @overload
    @staticmethod
    def create_dense_complex(
        type: RankedStructure[ComplexType[_FloatAttrTypeInvT]], data: Sequence[tuple[float, float]]
    ) -> DenseIntOrFPElementsAttr[ComplexType[_FloatAttrTypeInvT]]: ...

    @overload
    @staticmethod
    def create_dense_complex(
        type: RankedStructure[ComplexType[ComplexElementT]], data: Sequence[tuple[float, float]] | Sequence[tuple[float, float]]
    ) -> DenseIntOrFPElementsAttr[ComplexType[ComplexElementT]]: ...

    @staticmethod
    def create_dense_complex(
        type: RankedStructure[ComplexType],
        data: Sequence[tuple[float, float]] | Sequence[tuple[int, int]],
    ) -> DenseIntOrFPElementsAttr[ComplexType]:
        return DenseIntOrFPElementsAttr([type, BytesAttr(type.element_type.pack(data))])

```
if you add a `TypeVar` `_IntegerTypeInvT`

Reply:
https://github.com/xdslproject/xdsl/pull/4379/commits/92cd7fc07f005d882897dbc862b8b607a2421513 Thank you! Do you recommend any resources for learning pyright? Thanks!

Reply:
I'll try to do a full review later, then hopefully we can move towards getting this merged.

Reply:
Thank you :)